pykern/pktype.py

```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
+"""Validated types
+
+:copyright: Copyright (c) 2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from __future__ import absolute_import, division, print_function
+from pykern.pkcollections import PKDict
+from pykern import pkconfig
+
+
+class PKType():
```

Comment:
The types themselves, since they are capitalized, don't need to be PKType. We will never refer to them unqualified. Therefore, this type should be pktype.Base. 

The larger issue is that we instances of these types so that we aren't constantly referring to the schema to type check a result or whatever. This gets pretty tricky. You would use [`__new__` ](https://docs.python.org/3/reference/datamodel.html#object.__new__) for immutable types (int, bool, str, frozenset, etc.). The `__new__` would take as its argument, say, a string, and return an instance of cls. If this failed, it would raise ValueError or TypeError, similar to what int, bool, etc. do.

To convert these values to strings, I think we might want to consider a separate class that knows how to dispatch conversions to other classes. If this is the case, it might be better to not subclass the immutable types, and instead use them directly. 

Another alternative is multiple inheritance, but I like that a lot less.

IOW, these needs a bit more thought and/or discussion.

Reply:
https://github.com/radiasoft/sirepo/issues/4167#issuecomment-1073254224 discusses pydantic. We should look at existing solutions to see if they do enough of what we want.

Reply:
The `typing` module might even suffice; I'm looking at that now. I didn't know about it until browsing pydantic.